### ISSUE TITLE
Training Time?

### ISSUE THREAD

**Issue by avdravid** [EIFY/mup-vit#5]
Hi, thank you for the great repo. I am running the main branch. You benchmarked the code and got that it takes slightly less than 6 hours on 8x A100-SXM4-40GB. I am running on 8x A100 80GB PCIe and getting around 15 hours. I am running with Torch 2.4.0 and CUDA 12.4. I am wondering what is causing this large discrepancy as it is hard to imagine that the difference between SXM and PCIe would cause such a difference. This 15 hours is also with further optimizations I implemented such as Flash Attention, Amp, Bfloat16.

**Comment by EIFY** [comment]
Hi @avdravid,

Thanks for filing the bug! While I have never tried a PCIe A100 instance, I agree that it shouldn't take 15 hours. I would say first revert your changes (Flash Attention, Amp, Bfloat16): [`torch.nn.MultiheadAttention`](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html) should be quite optimized already and training in Bfloat16 is unlikely to reach reasonable accuracy (see https://github.com/google-research/vision_transformer/issues/153 & https://github.com/google-research/big_vision/issues/41). After that, perhaps consider upgrading to Torch 2.5.1 that I have been running.

If it still takes around 15 hours with all of the above then we may need to dig deeper.

**Comment by avdravid** [comment]
Thanks for the reply. I will try that and get back to you.

**Comment by avdravid** [comment]
Hi @EIFY. I reverted the code and upgraded to Torch 2.5.1 with CUDA 12.4. It is still estimated to take a while. I've attached a screenshot of the logs. 

<img width="890" alt="Image" src="https://github.com/user-attachments/assets/7363f781-8865-4797-bd88-8e705a44c923" />

**Comment by avdravid** [comment]
> Hi [@EIFY](https://github.com/EIFY). I reverted the code and upgraded to Torch 2.5.1 with CUDA 12.4. It is still estimated to take a while. I've attached a screenshot of the logs.
> 
> <img alt="Image" width="890" src="https://private-user-images.githubusercontent.com/46203730/422603963-7363f781-8865-4797-bd88-8e705a44c923.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE5MTg2MzUsIm5iZiI6MTc0MTkxODMzNSwicGF0aCI6Ii80NjIwMzczMC80MjI2MDM5NjMtNzM2M2Y3ODEtODg2NS00Nzk3LWJkODgtOGU3MDVhNDRjOTIzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE0VDAyMTIxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJkMjgyMzlmYTUyZGYxZGZjMDUwM2NjZTM4MDk0MmViZjk3ZGUyZmQxNGFmMTkyOTU3MDUyYTBhNTYyOWY5YzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.L38fEAI_ZW91OlxjfYni15uvq81uYsKnrE6GcqKJPBU">

This is the command I am running: 

PYTHON=torchrun
N_WORKERS=80
N_THREADS=124


NUMEXPR_MAX_THREADS=$N_THREADS $PYTHON main.py /home/amil/ilsvrc_2024-01-04_1913 \
    --workers $N_WORKERS \
    --multiprocessing-distributed \
    --epochs 90 \
    --batch-size 1024\
    --torchvision-inception-crop\

**Comment by EIFY** [comment]
If it's still running I kind of want to see more converged ETA later. When I run on a 8x A100-SXM4 (40GB or 80GB) it converges faster but things could be different on your side.

A few other quick things you may want to check: run `nvidia-smi` and see if some of the GPUs are not drawing much power or have consistently low GPU-Util. If you use wandb then the "System" tab should also have some info.

**Comment by avdravid** [comment]
Thanks. I'll try that and get back to you.

**Comment by avdravid** [comment]
Here is a report of the run so far: https://api.wandb.ai/links/amil/4qz6bjed. And the logs below. 

<img width="890" alt="Image" src="https://github.com/user-attachments/assets/adce047f-17b5-47fc-85b0-e28dd7b85c92" />

**Comment by EIFY** [comment]
Alright so I would say it doesn't look too unreasonable now and ETA should converge even lower. Depending on the details of your instance though, you may want to tune `N_WORKERS` and `N_THREADS` below even further:

```
PYTHON=torchrun
N_WORKERS=80
N_THREADS=124

NUMEXPR_MAX_THREADS=$N_THREADS $PYTHON main.py /home/amil/ilsvrc_2024-01-04_1913
--workers $N_WORKERS
--multiprocessing-distributed
--epochs 90
--batch-size 1024
--torchvision-inception-crop\
```
`N_THREADS` can be up to the number of logical threads on your instance, maybe just leave out a few for system bookkeeping. `N_WORKERS` can be up to `N_THREADS - 8` (minus one for each GPU) but it depends whether that's optimal. If it goes well it could reduce "Data" time or variance but you might be hitting diminishing return. Lastly, `--prefetch-factor 2` or any values up to `8` might be worth trying*.

(*There is some subtlety regarding the number of distinct random values of mixup lambda over the batches, but I highly doubt that it matters)

**Comment by avdravid** [comment]
> Alright so I would say it doesn't look too unreasonable now and ETA should converge even lower. Depending on the details of your instance though, you may want to tune `N_WORKERS` and `N_THREADS` below even further:
> 
> ```
> PYTHON=torchrun
> N_WORKERS=80
> N_THREADS=124
> 
> NUMEXPR_MAX_THREADS=$N_THREADS $PYTHON main.py /home/amil/ilsvrc_2024-01-04_1913
> --workers $N_WORKERS
> --multiprocessing-distributed
> --epochs 90
> --batch-size 1024
> --torchvision-inception-crop\
> ```
> 
> `N_THREADS` can be up to the number of logical threads on your instance, maybe just leave out a few for system bookkeeping. `N_WORKERS` can be up to `N_THREADS - 8` (minus one for each GPU) but it depends whether that's optimal. If it goes well it could reduce "Data" time or variance but you might be hitting diminishing return. Lastly, `--prefetch-factor 2` or any values up to `8` might be worth trying*.
> 
> (*There is some subtlety regarding the number of distinct random values of mixup lambda over the batches, but I highly doubt that it matters)

Thanks! I appreciate it!
